src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs

```diff
@@ -3137,5 +3132,346 @@ static void M()
+    { }
+}
+class AttributeNamedArgument : Attribute
+{
+    static int Field = 0;
+    public string Instance;
+    [AttributeNamedArgument("""", Instance = nameof(Field))]
+    AttributeNamedArgument(string param)
+    { }
+}
+class GenericClass<T>
+{
+    public static T Field = default(T);
+}
+class Program
+{
+    static void RefGenericField(string param1 = nameof(GenericClass<int>.Field))
+    { }
+
+    static int F1var = 2;
+    static void F1(string a = nameof(F1var)) { }
+
+    static int F2var = 2;
+    int this[int dummy, string a = nameof(F2var)] { set { } }
+
+    static int F3var = 2;
+    static void AttributeInParam([AttributeArgument(nameof(F3var))] string s) { }
+    static int F4var = 2;
+    static void NamedAttributeInParam([AttributeArgument(param: nameof(F4var))] string s) { }
+    static int F5var = 2;
+    static void NamedAttributeInParam2([AttributeArgument(""2"", Instance = nameof(F5var))] string s) { }
+
+    static void Main()
+    {
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib46(source, parseOptions: DefaultParseOptions, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(16821, "https://github.com/dotnet/roslyn/issues/16821")]
+        public void SpeculativeDoesNotMarkUsed()
+        {
+            var source = SyntaxFactory.ParseSyntaxTree(@"
+class Program
+{
+    private static int Unused;
+    static void Main()
+    {
+    }
+}
+class A : System.Attribute
+{
+    public A(int x) { }
+}
+");
+            var speculative = SyntaxFactory.ParseSyntaxTree(@"
+class C
+{
+    [A(Program.Unused)]
+    public void M()
+    {
+    }
+}");
+            var comp = CreateStandardCompilation(source);
+
+            var semanticModel = comp.GetSemanticModel(source);
+
+            var position = source.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Single().Position;
+
+            var x = speculative.GetRoot().DescendantNodes().OfType<AttributeSyntax>().Single();
+            var attributeConstructor = (SourceConstructorSymbol)semanticModel.GetSpeculativeSymbolInfo(position, x).Symbol;
+            Assert.Equal("A", attributeConstructor.ContainingType.Name);
+
+            Assert.True(semanticModel.TryGetSpeculativeSemanticModel(position, x, out var speculativeModel));
+            var attributeConstructor2 = speculativeModel.GetSymbolInfo(x).Symbol;
+            Assert.Equal("A", attributeConstructor2.ContainingType.Name);
+
+            comp.VerifyDiagnostics(
+                // (4,24): warning CS0169: The field 'Program.Unused' is never used
+                //     private static int Unused;
+                Diagnostic(ErrorCode.WRN_UnreferencedField, "Unused").WithArguments("Program.Unused").WithLocation(4, 24)
+                );
+        }
+
+        [Fact]
+        [WorkItem(16821, "https://github.com/dotnet/roslyn/issues/16821")]
+        public void AttributesMarkUsed()
```

Comment:
>AttributesMarkUsed [](start = 20, length = 18)

We should test similar scenarios when attributes reference fields rather than locals.